portability-transfer/src/main/java/org/dataportabilityproject/transfer/PortabilityInMemoryDataCopier.java

```diff
@@ -121,16 +121,69 @@ private void copyHelper(
             exportAuthData,
             importAuthData,
             Optional.of(new ExportInformation(
-                continuationData.getPaginationData(), exportInformation.get().getContainerResource())));
+                continuationData.getPaginationData(),
+                exportInformation.get().getContainerResource())));
       }
 
       // Start processing sub-resources
       if (continuationData.getContainerResources() != null
           && !continuationData.getContainerResources().isEmpty()) {
         for (ContainerResource resource : continuationData.getContainerResources()) {
-          copyHelper(jobId, exportAuthData, importAuthData, Optional.of(new ExportInformation(null, resource)));
+          copyHelper(jobId, exportAuthData, importAuthData,
+              Optional.of(new ExportInformation(null, resource)));
+        }
+      }
+    }
+  }
+
+  private ExportResult runExportLogic(UUID jobId, AuthData exportAuthData,
+      Optional<ExportInformation> exportInformation) {
+    ExportResult<?> exportResult = exportHelper(jobId, exportAuthData, exportInformation);
+    if (exportResult.getType() == ResultType.ERROR) {
+      ExceptionResponse response = checkRetry(exportResult.getMessage());
+      if (response.canRetry) {
+        int attempts = 1;
+        // TODO: what if different kinds of errors have different max retries, and we start with one
+        // kind of error and then start seeing another?
+        while (attempts < response.maxRetries && response.canRetry) {
+          exportResult = exportHelper(jobId, exportAuthData, exportInformation);
+          if (exportResult.getType() != ResultType.ERROR) {
+            return exportResult;
+          } else {
+            attempts += 1;
+            response = checkRetry(exportResult.getMessage());
+          }
         }
       }
     }
+    return exportResult;
+  }
+
+  private ExportResult exportHelper(UUID jobId, AuthData exportAuthData,
+      Optional<ExportInformation> exportInformation) {
+    logger.debug("Starting export");
```

Comment:
if you're going to log, I think it would be more useful to include some more info